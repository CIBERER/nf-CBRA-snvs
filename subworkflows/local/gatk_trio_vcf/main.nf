include { GATK4_HAPLOTYPECALLER                                          }      from '../../../modules/nf-core/gatk4/haplotypecaller/main'

include { BCFTOOLS_SORT                                                  }      from '../../../modules/nf-core/bcftools/sort/main'
include { GATK4_MERGEVCFS                                                }      from '../../../modules/nf-core/gatk4/mergevcfs/main'

include { TABIX_TABIX                                                    }      from '../../../modules/nf-core/tabix/tabix/main'
include { BCFTOOLS_FILTER                                                }      from '../../../modules/nf-core/bcftools/filter/main'
include { SPLITMULTIALLELIC                                                }      from '../../../modules/local/splitmultiallelic/main'

include { GATK4_SELECTVARIANTS  as  GATK4_SELECTVARIANTS_SNP             }      from '../../../modules/nf-core/gatk4/selectvariants/main'
include { GATK4_SELECTVARIANTS  as  GATK4_SELECTVARIANTS_INDEL           }      from '../../../modules/nf-core/gatk4/selectvariants/main'
include { GATK4_SELECTVARIANTS  as  GATK4_SELECTVARIANTS_MIX             }      from '../../../modules/nf-core/gatk4/selectvariants/main'
include { GATK4_VARIANTFILTRATION  as  GATK4_VARIANTFILTRATION_SNV       }      from '../../../modules/nf-core/gatk4/variantfiltration/main'
include { GATK4_VARIANTFILTRATION  as  GATK4_VARIANTFILTRATION_INDEL     }      from '../../../modules/nf-core/gatk4/variantfiltration/main'
include { GATK4_VARIANTFILTRATION  as  GATK4_VARIANTFILTRATION_MIX       }      from '../../../modules/nf-core/gatk4/variantfiltration/main'
include { GATK4_VARIANTFILTRATION  as  GATK4_VARIANTFILTRATION_GENOTYPEPOSTERIOR       }      from '../../../modules/nf-core/gatk4/variantfiltration/main'

include { GATK4_GENOMICSDBIMPORT } from '../../../modules/nf-core/gatk4/genomicsdbimport/main'
include { GATK4_GENOTYPEGVCFS } from '../../../modules/nf-core/gatk4/genotypegvcfs/main'
include { GATK4_CALCULATEGENOTYPEPOSTERIORS } from '../../../modules/local/gatk4/calculategenotypeposteriors/main'

include { GATK4_VARIANTRECALIBRATOR as VARIANTRECALIBRATOR_INDEL} from '../../../modules/nf-core/gatk4/variantrecalibrator/main'
include { GATK4_VARIANTRECALIBRATOR as VARIANTRECALIBRATOR_SNP} from '../../../modules/nf-core/gatk4/variantrecalibrator/main'

// TODO: Create a module for CalculateGenotypePosteriors
// TODO: Create a module for ANNOTATEVARIANTS


workflow GATK_TRIO_VCF {

    take:
    ch_bam        // channel (mandatory): [ val(meta), path(bam), path(bai) ]
    ch_intervals    // channel (mandatory) : [ val(meta), path(bed) ]
    ch_fasta        // channel (mandatory) : [ val(meta2), path(fasta) ]
    ch_fai          // channel (mandatory) : [ val(meta2), path(fai) ]
    ch_refdict      // channel (mandatory) : [ val(meta2), path(dict) ]
    ch_dbsnp  // channel (mandatory) : [ val(meta3), path(vcf) ]
    ch_dbsnp_tbi  // channel (mandatory) : [ val(meta3), path(vcf) ]
    ch_intervals_genomicsdbimport
    no_intervals // channel (mandatory) : [ boolean ]
    ch_ped

    main:

    ch_versions = Channel.empty()
    
    ch_dragstr_model = ch_bam.map {meta, bam, bai -> tuple(meta, [])}


    GATK4_HAPLOTYPECALLER (
        ch_bam.join(ch_intervals).join(ch_dragstr_model),
        ch_fasta,
        ch_fai,
        ch_refdict,
        ch_dbsnp,
        ch_dbsnp_tbi
    )
    ch_versions = ch_versions.mix(GATK4_HAPLOTYPECALLER.out.versions.first())

    // Collect all VCFs and TBIs from all samples
    // First, prepare your joint VCF data
    joint_vcf_data = GATK4_HAPLOTYPECALLER.out.vcf
        .join(GATK4_HAPLOTYPECALLER.out.tbi, by: 0)//.view()
        .map { meta, vcf, tbi ->
        // Extract family from meta and create new tuple with family as key
        def family = meta.family
        tuple(family, meta, vcf, tbi)
        }
        .groupTuple(by: 0)
        .map { family, metas, vcfs, tbis ->
        // Create joint metadata for the family
        def joint_meta = [
            id: family,
            samples: metas.collect { it.id },
            sample_count: metas.size()
        ]
        tuple(joint_meta, vcfs, tbis)
    }//.view()

    // Then combine with intervals
    joint_gvcf_ch = joint_vcf_data
        .combine(ch_intervals_genomicsdbimport)
        .map { joint_meta, vcfs, tbis, intervals ->
            tuple(joint_meta, vcfs, tbis, intervals, [], [])
    }//.view()

    GATK4_GENOMICSDBIMPORT(joint_gvcf_ch, false, false, false)

    if (no_intervals) {
        // If no intervals are provided, we can use the whole genome
        genotype_input = GATK4_GENOMICSDBIMPORT.out.genomicsdb.map{ meta, genomicsdb -> [ meta, genomicsdb, [], [], [] ] }//.view { "NO BED: $it" }
    } else {
        // Use the provided intervals
        genotype_input = GATK4_GENOMICSDBIMPORT.out.genomicsdb.combine(ch_intervals.map { meta, bed -> bed }.first()).map{ meta, genomicsdb, intervals -> [ meta, genomicsdb, [], intervals, [] ] }//.view { "WITH BED: $it" }
    }
    
    //genotype_input.view()

    GATK4_GENOTYPEGVCFS (
        genotype_input,
        ch_fasta,
        ch_fai,
        ch_refdict,
        ch_dbsnp,
        ch_dbsnp_tbi
    )
    
    GATK4_GENOTYPEGVCFS.out.vcf.join(GATK4_GENOTYPEGVCFS.out.tbi).view { "GATK4_GENOTYPEGVCFS.out.vcf: $it" }

    if (no_intervals) {
        // If no intervals are provided, we can use the whole genome
        ch_for_selectvariants = GATK4_GENOTYPEGVCFS.out.vcf
        .join(GATK4_GENOTYPEGVCFS.out.tbi).map { meta, vcf, tbi -> [meta, vcf, tbi, []]
        }
    } else {
        // Use the provided intervals
        ch_for_selectvariants = GATK4_GENOTYPEGVCFS.out.vcf
        .join(GATK4_GENOTYPEGVCFS.out.tbi)
        .combine(ch_intervals.map { meta, bed -> bed }.first())
    }

    ch_for_selectvariants.view()

    // TODO: create subworkflows for HardFiltering

    GATK4_SELECTVARIANTS_SNP (
        ch_for_selectvariants
    )
    ch_versions = ch_versions.mix(GATK4_SELECTVARIANTS_SNP.out.versions.first())

    GATK4_SELECTVARIANTS_INDEL(
        ch_for_selectvariants
    )
    ch_versions = ch_versions.mix(GATK4_SELECTVARIANTS_INDEL.out.versions.first())

    GATK4_SELECTVARIANTS_MIX(
        ch_for_selectvariants
    )
    ch_versions = ch_versions.mix(GATK4_SELECTVARIANTS_MIX.out.versions.first())

    GATK4_VARIANTFILTRATION_SNV(
        GATK4_SELECTVARIANTS_SNP.out.vcf.join(GATK4_SELECTVARIANTS_SNP.out.tbi),
        ch_fasta,
        ch_fai,
        ch_refdict,
    ) 
    ch_versions = ch_versions.mix(GATK4_VARIANTFILTRATION_SNV.out.versions.first())

    GATK4_VARIANTFILTRATION_INDEL(
        GATK4_SELECTVARIANTS_INDEL.out.vcf.join(GATK4_SELECTVARIANTS_INDEL.out.tbi),
        ch_fasta,
        ch_fai,
        ch_refdict,
    )
    ch_versions = ch_versions.mix(GATK4_VARIANTFILTRATION_INDEL.out.versions.first())

    GATK4_VARIANTFILTRATION_MIX(
        GATK4_SELECTVARIANTS_MIX.out.vcf.join(GATK4_SELECTVARIANTS_MIX.out.tbi),
        ch_fasta,
        ch_fai,
        ch_refdict,
    )
    ch_versions = ch_versions.mix(GATK4_VARIANTFILTRATION_MIX.out.versions.first())

    BCFTOOLS_SORT(
        GATK4_VARIANTFILTRATION_SNV.out.vcf.concat( GATK4_VARIANTFILTRATION_INDEL.out.vcf, GATK4_VARIANTFILTRATION_MIX.out.vcf )
    )
    ch_versions = ch_versions.mix(BCFTOOLS_SORT.out.versions.first())

    GATK4_MERGEVCFS(
        BCFTOOLS_SORT.out.vcf.groupTuple(),
        ch_refdict
    )
    ch_versions = ch_versions.mix(GATK4_MERGEVCFS.out.versions.first())

    BCFTOOLS_FILTER(
        GATK4_MERGEVCFS.out.vcf.join(GATK4_MERGEVCFS.out.tbi)
    )
    ch_versions = ch_versions.mix(BCFTOOLS_FILTER.out.versions.first())

    BCFTOOLS_FILTER.out.vcf.view()

    // TODO: add a module for CalculateGenotypePosteriors
    // TODO: add a module for ANNOTATEVARIANTS


    if (no_intervals) {
        // If no intervals are provided, we can use the whole genome
        ch_input_genotypeposteriors = BCFTOOLS_FILTER.out.vcf.join(BCFTOOLS_FILTER.out.tbi).join(ch_ped).map { meta, vcf, ped -> [meta, vcf, [], ped]}
    } else {
        // Use the provided intervals
        ch_input_genotypeposteriors = BCFTOOLS_FILTER.out.vcf.join(BCFTOOLS_FILTER.out.tbi).join(ch_ped).combine(ch_intervals.map { meta, bed -> bed }.first())
        .map { meta, vcf, ped, interval -> [meta, vcf, interval, ped]}
    }

    GATK4_CALCULATEGENOTYPEPOSTERIORS(
        ch_input_genotypeposteriors
    )




    vcf = GATK4_HAPLOTYPECALLER.out.vcf.join(GATK4_HAPLOTYPECALLER.out.tbi)//.view()

    emit:
    vcf // channel: [ val(meta), path(vcf), path(tbi)]
    versions = ch_versions                     // channel: [ versions.yml ]

}