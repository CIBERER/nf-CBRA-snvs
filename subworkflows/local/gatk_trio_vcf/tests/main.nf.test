nextflow_workflow {

    name "Test Subworkflow GATK_TRIO_VCF"
    script "../main.nf"
    workflow "GATK_TRIO_VCF"

    tag "subworkflows"
    tag "subworkflows/gatk_vcf"
    tag "subworkflows/mapping"
    tag "bwa"
    tag "bwa/mem"
    tag "gatk4"
    tag "gatk4/fastqtosam"
    tag "gatk4/mergebamalignment"
    tag "gatk4/fastqtosam"
    tag "gatk4/baserecalibrator"
    tag "gatk4/applybqr"
    tag "gatk4spark"
    tag "gatk4spark/markduplicates"    
    tag "picard"
    tag "picard/sortsam"

    setup {

        run("BWA_INDEX") {
            script "../../../../modules/nf-core/bwa/index/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        
        run("PICARD_CREATESEQUENCEDICTIONARY") {
            script "../../../../modules/nf-core/picard/createsequencedictionary/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        
        // Run mapping for all three samples at once
        run("MAPPING") {
            script "../../../../subworkflows/local/mapping/main.nf"
            process {
                """
                // Create channel with three samples
                input[0] = Channel.of(
                    [
                        [ id : 'sample1_father' ],
                        [file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_2.fastq.gz', checkIfExists: true)]
                    ],
                    [
                        [ id : 'sample2_mother' ],
                        [file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_2.fastq.gz', checkIfExists: true)]
                    ],
                    [
                        [ id : 'sample3_child' ],
                        [file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_2.fastq.gz', checkIfExists: true)]
                    ]
                )
                
                input[1] = Channel.of(
                    tuple([ id: 'sample1_father'], []),
                    tuple([ id: 'sample2_mother'], []),
                    tuple([ id: 'sample3_child'], [])
                )
                
                input[2] = BWA_INDEX.out.index
                
                input[3] = Channel.fromList([
                    tuple([ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                ]).collect()
                
                input[4] = Channel.fromList([
                    tuple([ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                ]).collect()
                
                input[5] = PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict
                
                input[6] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz', checkIfExists: true)
                
                input[7] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi', checkIfExists: true)
                """
            }
        }
    }

    test("gatk_trio_vcf - three samples") {

        when {
            workflow {
                """
                // Use the BAM output from mapping (should contain all three samples)
                input[0] = MAPPING.out.bam
                
                // Intervals channel - empty intervals for all samples
                input[1] = MAPPING.out.bam.map { meta, bam, bai -> 
                    tuple(meta, [])
                }
                
                // Reference genome
                input[2] = Channel.fromList([
                    tuple([ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                ]).collect()
                
                // Reference genome index
                input[3] = Channel.fromList([
                    tuple([ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                ]).collect()
                
                // Reference dictionary
                input[4] = PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict
                
                // dbSNP VCF
                input[5] = Channel.fromList([
                    tuple([ id: 'test_dbsnp'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz', checkIfExists: true))
                ]).collect()
                
                // dbSNP TBI
                input[6] = Channel.fromList([
                    tuple([ id: 'test_dbsnp'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi', checkIfExists: true))
                ]).collect()

                // intervals_for_genomicsdbimport
                input[7] = Channel.fromPath(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.interval_list', checkIfExists: true).collect()
                input[8] = "true"

                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}