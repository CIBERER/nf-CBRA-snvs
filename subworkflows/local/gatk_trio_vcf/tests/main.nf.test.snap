{
    "gatk_trio_vcf - three samples": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "joint_variant_calling",
                            "samples": [
                                "sample2_mother",
                                "sample3_child",
                                "sample1_father"
                            ],
                            "sample_count": 3
                        },
                        "joint_variant_calling.gatk.PASS.vcf.gz:md5,54630fd14a439c50f41baaae228c08f8",
                        "joint_variant_calling.gatk.PASS.vcf.gz.tbi:md5,6dba91748febecab0371836408ff5ac8"
                    ]
                ],
                "1": [
                    "versions.yml:md5,05bec2fa7388dd27228b4bd1ad4f34bf",
                    "versions.yml:md5,1582383cb46789aff840ab7b4d010a73",
                    "versions.yml:md5,1fce0729b06abba7a681e153298a444c",
                    "versions.yml:md5,2d1c68ffdb67ba71ca82db345e5bcb0e",
                    "versions.yml:md5,3206879867c823204ab00fcbbed32714",
                    "versions.yml:md5,4ff24260bb047ac6f911721e74f1701d",
                    "versions.yml:md5,66749257146ead09e960eff355164d78",
                    "versions.yml:md5,92700e62d2a10d20c01fcd62963b077a",
                    "versions.yml:md5,e6b4ce3ffdb921b48fd81f502caf0cb9",
                    "versions.yml:md5,e744d60cfdf265fc59a006775ac6a8b9"
                ],
                "vcf": [
                    [
                        {
                            "id": "joint_variant_calling",
                            "samples": [
                                "sample2_mother",
                                "sample3_child",
                                "sample1_father"
                            ],
                            "sample_count": 3
                        },
                        "joint_variant_calling.gatk.PASS.vcf.gz:md5,54630fd14a439c50f41baaae228c08f8",
                        "joint_variant_calling.gatk.PASS.vcf.gz.tbi:md5,6dba91748febecab0371836408ff5ac8"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,05bec2fa7388dd27228b4bd1ad4f34bf",
                    "versions.yml:md5,1582383cb46789aff840ab7b4d010a73",
                    "versions.yml:md5,1fce0729b06abba7a681e153298a444c",
                    "versions.yml:md5,2d1c68ffdb67ba71ca82db345e5bcb0e",
                    "versions.yml:md5,3206879867c823204ab00fcbbed32714",
                    "versions.yml:md5,4ff24260bb047ac6f911721e74f1701d",
                    "versions.yml:md5,66749257146ead09e960eff355164d78",
                    "versions.yml:md5,92700e62d2a10d20c01fcd62963b077a",
                    "versions.yml:md5,e6b4ce3ffdb921b48fd81f502caf0cb9",
                    "versions.yml:md5,e744d60cfdf265fc59a006775ac6a8b9"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-07-17T09:36:35.259508145"
    }
}